### PR TITLE
Integration tests: /router/quote

### PR DESCRIPTION
**Pull request by p0mvn**
https://linear.app/osmosis/issue/DATA-61/[sqs-automated-tests-in-ci][quotes]

- Added quote tests
- Refactored test setup logic to ensure determinism when running in-parallel
- Added documentation in `tests/README.md` **Please read it before reviewing the tests**

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -19,3 +32,489 @@ def environment_url(request):
     SQS_STAGE: SERVICE_SQS_STAGE,
     SQS_PROD: SERVICE_SQS_PROD
 }
+
+# Numia pool type constants using an Enum
+class NumiaPoolType(Enum):
```
Comment (p0mvn) [inline tests/conftest.py:79]: Note to reviewer: most of the diff in this file was moved from `setup.py` due to refactoring setup logic and having to rely on the built-in functionality of `conftest.py`.

I will mark the new areas of the codebase with comments

File: tests/conftest.py [hunk 2]
```diff
@@ -19,3 +32,503 @@ def environment_url(request):
...
+
+    # Format pairs for return
+    formatted_pairs = [[token1, token2] for token1, token2 in token_pairs]
+
+    if len(formatted_pairs) > MIN_NUM_MISC_TOKEN_PAIRS:
+        ValueError(f"Constructeed {len(formatted_pairs)}, min expected {MIN_NUM_MISC_TOKEN_PAIRS}")
+
+    return formatted_pairs
+
+def create_no_dupl_token_pairs(token_list):
```
Comment (p0mvn) [inline tests/conftest.py:452]: Note to reviewer: the methods starting from this one are new and should be reviewed in greater detail

File: tests/README.md [hunk 3]
```diff
@@ -17,6 +17,13 @@ make e2e-install-requirements
 make e2e-run-dev
 ```
 
+### Isolate Specific Test
+
+```bash
+# Run a specific test from master node (one process)
+ pytest -s tests/test_router_quote.py::TestQuote::test_usdc_in_high_liq_out[https://sqs.stage.osmosis.zone-62152071397ibc/69110FF673D70B39904FF056CFDFD58A90BEC3194303F45C32CB91B8B0A738EA]
```
Comment (cryptomatictrader) [inline tests/README.md:24]: Nice!

File: tests/README.md [hunk 4]
```diff
@@ -32,3 +39,45 @@ We have a debug configuration under `.vscode/launch.json` called .
...
 It alows setting breakpoints for streamlined debugging.
+
+## Test Suite Setup
+
+There are 2 modes of running the suite:
+1. Single-process
+2. Multi-process
+
+With the multi-process mode, the setup logic in `conftest.py` is executed by every worker.
+Every worker then aggregates their results to construct parametrized tests and split them across each other.
```
Comment (cryptomatictrader) [inline tests/README.md]: I'm sorry I don't quite follow the last sentence "Every worker then aggregates...". Could you please explain a bit more?

File: tests/README.md [hunk 5]
```diff
@@ -32,3 +39,45 @@ We have a debug configuration under `.vscode/launch.json` called .
...
+To prevent that, we run the setup logic to generate common test parameters from a master process. We write
+the output to a file. See `conftest.py::pytest_sessionstart` for more details.
+
+We serialize the setup state as `conftest.py::SharedTestState`, letting each worker to then deserialize it
+for deterministic test parameter generation.
+
+## Quote Test Suite
+
+We have 3 kinds of quotes tests:
+1. USDC input from 10^6 to 10^9 amount with every supported token as output denom
```
Comment (cryptomatictrader) [inline tests/README.md]: Do you mean USDC input with value min. 10^6 and max. 10^9? Or something to do with exponent?

File: tests/README.md [hunk 6]
```diff
@@ -32,3 +39,45 @@ We have a debug configuration under `.vscode/launch.json` called .
...
+2. Top X by-liquidity tokens with each other. Generate amounts from 10^6 to 10^9
```
Comment (cryptomatictrader) [inline tests/README.md]: Same question around 10^6 and 10^9. Appreciate if a simple example is given.

File: tests/conftest.py [hunk 7]
```diff
@@ -19,3 +32,506 @@ def environment_url(request):
...
+    """
+    Creates all unique combinations of token pairs from a list of tokens.
```
Comment (cryptomatictrader) [inline tests/conftest.py:454]: IS ATOM-USDC and USDC-ATOM considered different? I guess so.

File: tests/conftest.py [hunk 8]
```diff
@@ -19,3 +32,506 @@ def environment_url(request):
...
+    filtered_list = []
+    for combo in combinations:
+        if combo[0] != combo[1]:
+            filtered_list.append(combo)
+
+    return filtered_list
+
+def create_coins_from_pairs(pairs, start_order, end_order):
+    """
+    Create a list of token in and out pairs from the given list of token pairs, starting and ending orders of magnitude.
```
Comment (cryptomatictrader) [inline tests/conftest.py:470]: What is meant by "starting and ending orders of magnitude" (sorry for lack of context, I just imagine reading it from a fresh eyes). Appreciate if there is a simple example of parameter values => returned values.

File: tests/conftest.py [hunk 9]
```diff
@@ -19,3 +32,506 @@ def environment_url(request):
...
+    denom_data = shared_test_state.chain_denom_to_data_map.get(denom)
+    return denom_data.get("exponent")
+
+def pytest_sessionstart(session):
+    """
+    This hook is called after the Session object has been created and
+    before performing collection and entering the run test loop.
+
+    This is where we perform the setup tasks for the tests.
+    If the code is running on the master node, we fetch all the data once and store it in a shared state.
```
Comment (cryptomatictrader) [inline tests/conftest.py]: "fetch all the data" => where does the data come from? I see the `fetch_tokens()` & `fetch_pools()` in the code and wonder where it is from

File: tests/conftest.py [hunk 10]
```diff
@@ -19,3 +32,506 @@ def environment_url(request):
...
+
+        # One Transmuter token pair [[pool_id, ['denom0', 'denom1']]]
+        shared_test_state.transmuter_token_pairs = choose_transmuter_pool_tokens_by_liq_asc(shared_test_state.pool_type_to_denoms, 1)
+
+        # One Astroport token pair [[pool_id, ['denom0', 'denom1']]]
+        shared_test_state.astroport_token_pair = choose_pcl_pool_tokens_by_liq_asc(shared_test_state.pool_type_to_denoms, 1)
+
+        shared_test_state.misc_token_pairs = create_misc_token_pairs()
+
+        with FileLock(sqs_e2e_data_lock_file):
```
Comment (cryptomatictrader) [inline tests/conftest.py:573]: 👍

File: tests/conftest.py [hunk 11]
```diff
@@ -19,3 +32,506 @@ def environment_url(request):
...
+    state file for determinism.
+
+    See tests/README.md for details.
+    """
+    print("Session is starting. Worker ID:", getattr(session.config, 'workerinput', {}).get('workerid', 'master'))
+
+    global shared_test_state
+
+    # Example setup logic
+    if not hasattr(session.config, 'workerinput'):  # This checks if the code is running on the master node
```
Comment (cryptomatictrader) [inline tests/conftest.py:540]: Below block looks good and clear to me 👍

File: tests/sqs_service.py [hunk 12]
```diff
@@ -1,9 +1,10 @@
 import requests
 
-SQS_STAGE = "https://sqs.stage.osmosis.zone"
+SQS_STAGE = "https://sqs-eu.stage.osmosis.zone"
```
Comment (cryptomatictrader) [inline tests/sqs_service.py]: Just curious - why specifically pointing to eu ?

File: tests/test_router_quote.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+import conftest
+from sqs_service import *
+from quote_response import *
+from rand_util import *
+from e2e_math import *
+from decimal import *
+from constants import *
+from util import *
+
+SQS_STAGE = "https://sqs.stage.osmosis.zone"
```
Comment (cryptomatictrader) [inline tests/test_router_quote.py]: Looks like we have multiple places in tests suite pointing to https://sqs.stage.osmosis.zone, /router/quotes and /router/routes. Shall we move them to constants.py? Same for QUOTE_NUM_TOP_LIQUIDITY_DENOMS, TRANSMUTER_MIN_TOKEN_LIQ_USD, expected_latency_upper_bound_ms, like what we did in test_tokens_prices.pay. Hopefully we don't have cyclic dependency issue and if so, it's okay to make them stay here.

File: tests/test_router_quote.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+        The amounts are constructed to be seeded random values between 10^USDC_PRECISION-1 and 10 ^(USDC_PRECISION + 4)
+
+        This allows us to validate that we can continue to quote at reasonable USDC values for all majore token pairs without errors.
+
+        Note: the reason we use Decimal in this test is because floats truncate in some edge cases, leading
+        to flakiness.
+        """
+        # This is the max error tolerance of 7% that we allow.
+        # Arbitrarily hand-picked to avoid flakiness.
+        error_tolerance = 0.07
```
Comment (cryptomatictrader) [inline tests/test_router_quote.py:35]: How about `error_tolerance` move to constants.py ?

File: tests/test_router_quote.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+    def test_top_liq_combos_default_exponent(self, environment_url, swap_pair):
+        token_in_obj = swap_pair['token_in']
+        amount_str = token_in_obj['amount_str']
+        token_in_denom = token_in_obj['denom']
+        token_in_coin = amount_str + token_in_denom
+        denom_out = swap_pair['out_denom']
+
+        # This is the max error tolerance of 7% that we allow.
+        # Arbitrarily hand-picked to avoid flakiness.
+        error_tolerance = 0.07
```
Comment (cryptomatictrader) [inline tests/test_router_quote.py:76]: Same

File: tests/test_router_quote.py [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+        skip_imbalanced_pool_test_if_imbalanced(transmuter_token_data)
+
+        transmuter_token_pair = transmuter_token_data[1]
+
+        denom_in = transmuter_token_pair[0]
+        denom_out = transmuter_token_pair[1]
+
+        # This is the max error tolerance of 5% that we allow.
+        # Arbitrarily hand-picked to avoid flakiness.
+        error_tolerance = 0.05
```
Comment (cryptomatictrader) [inline tests/test_router_quote.py:125]: Same

**Review by cryptomatictrader — COMMENTED** [review]
Great job! Especially around the master / worker mechanism and test preparation! I wish pytest has a better framework to support that! Thanks!

File: tests/test_router_quote.py (hunk 14 above)
Comment (p0mvn) [reply tests/test_router_quote.py:35]: This is by design. Each test benefits from having a custom error tolerance. They don't have a universal one shared across. Note that some of them already have different values.

We will continue tweaking the one for each based on the needs basis.

File: tests/test_router_quote.py (hunk 15 above)
Comment (p0mvn) [reply tests/test_router_quote.py:76]: https://github.com/osmosis-labs/sqs/pull/238#discussion_r1619730298

File: tests/test_router_quote.py (hunk 16 above)
Comment (p0mvn) [reply tests/test_router_quote.py:125]: https://github.com/osmosis-labs/sqs/pull/238#discussion_r1619730298

File: tests/README.md (hunk 4 above)
Comment (p0mvn) [reply tests/README.md]: Expanded here to enhance clarity:
[a287087](https://github.com/osmosis-labs/sqs/pull/238/commits/a287087f7f6d2ce8c2c9c7c71899e7eca08f3de7)

File: tests/README.md (hunk 5 above)
Comment (p0mvn) [reply tests/README.md]: Made it clearer in this commit:
[1452484](https://github.com/osmosis-labs/sqs/pull/238/commits/145248475d18a022c91bd2e6dddd7c420263b891)

File: tests/README.md (hunk 6 above)
Comment (p0mvn) [reply tests/README.md]: Fixed in this commit:
[50dc893](https://github.com/osmosis-labs/sqs/pull/238/commits/50dc8930936b3ef83b09143e17f3a8068b01c6f2)

File: tests/conftest.py (hunk 7 above)
Comment (p0mvn) [reply tests/conftest.py:454]: Same. These are combinations.

Permutations would be different

File: tests/sqs_service.py (hunk 12 above)
Comment (p0mvn) [reply tests/sqs_service.py]: Was debugging stage. Going to revert now - thanks for pointing out

File: tests/conftest.py (hunk 8 above)
Comment (p0mvn) [reply tests/conftest.py:470]: Example added: [0b860a8](https://github.com/osmosis-labs/sqs/pull/238/commits/0b860a838faaee832dff9c907e896a6be7856e6d)

File: tests/test_router_quote.py (hunk 13 above)
Comment (p0mvn) [reply tests/test_router_quote.py]: Removed duplicate stage - was an erroneous copy. Thanks.

Kept `TRANSMUTER_MIN_TOKEN_LIQ_USD` in `constants.py` since it is used across many test components.

On the other hand, `QUOTE_NUM_TOP_LIQUIDITY_DENOMS` is here for a reason. Instead of creating a bloated `constants.py` with identifiers that are only used in individual files, I prefer to define them in test files where they are used directly.

`constants.py` instead should contain constants that are shared across multiple files or components.

File: tests/conftest.py (hunk 9 above)
Comment (p0mvn) [reply tests/conftest.py]: Linked to `data_service.py`:
[88cc942](https://github.com/osmosis-labs/sqs/pull/238/commits/88cc942a049ee3ea03518c9a0c7fbf10dc3c2767)

It is Numia. But I don't think this is the right level of abstraction to specifically mention that. We might change that in the future.

If we do, this comment would be unlikely to be updated.

Instead, this low-level context is preserved in `data_service.py` and in the README.md while client components link them for details

**Comment by p0mvn** [comment]
@cryptomatictrader thanks for review. I addressed all comments

File: tests/conftest.py (hunk 9 above)
Comment (cryptomatictrader) [reply tests/conftest.py]: Got it and make sense. Thanks for clarifying that.

File: tests/test_router_quote.py (hunk 13 above)
Comment (cryptomatictrader) [reply tests/test_router_quote.py]: Make sense. Thannks!

File: tests/test_router_quote.py (hunk 14 above)
Comment (cryptomatictrader) [reply tests/test_router_quote.py:35]: OK. Let's keep it like that and we will revisit it as needed.

**Review by cryptomatictrader — APPROVED** [review]
Great work! And thanks for the explanation and clarification 👍
